File: packages/python-host/src/ai37_agent_host/store_backend/attachments_store_backend.py
"""HTTP-клиенты вложений — порт ``ts-host/src/store-backend/attachments-store-backend.ts``.

Тонкий httpx-клиент (async) к REST chat-backend. Файлы (Redis/Postgres/S3) знает только
chat-backend — единая точка auth/tenancy. Bearer форвардится из request-scope (``current_bearer``).

Расширение Фазы 4: ``ChatAttachmentsStoreBackend.read_raw`` отдаёт БАЙТЫ оригинала
(``GET /{fileId}/raw``) — для детерминированного парсинга (openpyxl). Base/project — read-only
(durable/S3 raw — follow-up).
"""

from __future__ import annotations

import re
from collections.abc import Callable
from typing import Any
from urllib.parse import quote

import httpx

from ..als import current_bearer
from .types import (
    EditResult,
    FileInfo,
    GlobResult,
    GrepMatch,
    GrepResult,
    LsResult,
    ReadRawResult,
    ReadResult,
    WriteResult,
)

_READ_ONLY = "Вложения доступны только для чтения (агенты read-only к file:<id>)"
_RAW_UNSUPPORTED = "readRaw не поддерживается: используйте read (markdown-текст)"

BearerFn = Callable[[], str | None]


class AttachmentsStoreBackendBase:
    """База StoreBackend'ов вложений: HTTP-клиент к chat-backend. Виртуальная ФС по ``anchor``."""

    anchor: str = ""
    api_base: str = ""

    def __init__(
        self,
        *,
        base_url: str,
        bearer: BearerFn | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._bearer = bearer or current_bearer
        self._client = http_client

    # ── абстрактные резолверы scope хода ──────────────────────────────────────
    def _scope_for_manifest(self) -> dict[str, str] | None:
        raise NotImplementedError

    def _scope_for_file(self) -> dict[str, str] | None:
        raise NotImplementedError

    # ── ls ────────────────────────────────────────────────────────────────────
    async def ls(self, path: str) -> LsResult:
        if self._parse(path) != "":
            return LsResult(error=f"Не директория: {path}")
        scope = self._scope_for_manifest()
        if scope is None:
            return LsResult(error=self._scope_missing())
        try:
            data = await self._api_json("/", scope)
            return LsResult(files=[self._file_info(a) for a in data.get("attachments", [])])
        except Exception as exc:  # noqa: BLE001
            return LsResult(error=str(exc))

    # ── read (директория → манифест; файл → окно markdown) ────────────────────
    async def read(
        self, file_path: str, offset: int | None = None, limit: int | None = None
    ) -> ReadResult:
        file_id = self._parse(file_path)
        if file_id is None:
            return ReadResult(error=f"Неизвестный путь: {file_path}")
        try:
            if not file_id:
                scope = self._scope_for_manifest()
                if scope is None:
                    return ReadResult(error=self._scope_missing())
                data = await self._api_json("/", scope)
                return ReadResult(
                    content=_render_manifest(self.anchor, data.get("attachments", [])),
                    mime_type="text/markdown",
                )
            scope = self._scope_for_file()
            if scope is None:
                return ReadResult(error=self._scope_missing())
            query = dict(scope)
            if offset is not None:
                query["offset"] = str(offset)
            if limit is not None:
                query["limit"] = str(limit)
            data = await self._api_json(f"/{quote(file_id)}/content", query)
            return ReadResult(content=data.get("content", ""), mime_type="text/markdown")
        except Exception as exc:  # noqa: BLE001
            return ReadResult(error=str(exc))

    # ── glob (по имени файла, клиентский фильтр манифеста) ─────────────────────
    async def glob(self, pattern: str, path: str | None = None) -> GlobResult:
        scope = self._scope_for_manifest()
        if scope is None:
            return GlobResult(error=self._scope_missing())
        try:
            needle = re.sub(r"[*?]", "", pattern).strip().lower()
            data = await self._api_json("/", scope)
            files = [
                self._file_info(a)
                for a in data.get("attachments", [])
                if not needle or needle in str(a.get("sourceName", "")).lower()
            ]
            return GlobResult(files=files)
        except Exception as exc:  # noqa: BLE001
            return GlobResult(error=str(exc))

    # ── grep (по содержимому, серверный поиск) ────────────────────────────────
    async def grep(
        self, pattern: str, path: str | None = None, glob: str | None = None
    ) -> GrepResult:
        scope = self._scope_for_manifest()
        if scope is None:
            return GrepResult(error=self._scope_missing())
        try:
            data = await self._api_json("/search", {**scope, "q": pattern})
            matches = [
                GrepMatch(
                    path=f"/{self.anchor}/{hit['fileId']}",
                    line=int(hit.get("line", 0)),
                    text=f"[{hit.get('sourceName', '')}] {_one_line(str(hit.get('snippet', '')))}",
                )
                for hit in data.get("matches", [])
            ]
            return GrepResult(matches=matches)
        except Exception as exc:  # noqa: BLE001
            return GrepResult(error=str(exc))

    # ── read-only / raw ───────────────────────────────────────────────────────
    async def write(self, file_path: str, content: str) -> WriteResult:
        return WriteResult(error=_READ_ONLY)

    async def edit(
        self, file_path: str, old_string: str, new_string: str, replace_all: bool = False
    ) -> EditResult:
        return EditResult(error=_READ_ONLY)

    async def read_raw(self, file_path: str) -> ReadRawResult:
        return ReadRawResult(error=_RAW_UNSUPPORTED)

    # ── helpers ────────────────────────────────────────────────────────────────
    def _parse(self, path: str) -> str | None:
        """fileId сегмента (str), '' для директории-якоря, None если якорь не найден."""
        segments = [s for s in path.split("/") if s]
        if self.anchor not in segments:
            return None
        index = segments.index(self.anchor)
        return segments[index + 1] if index + 1 < len(segments) else ""

    def _file_info(self, meta: dict[str, Any]) -> FileInfo:
        return FileInfo(
            path=f"/{self.anchor}/{meta.get('fileId')}",
            is_dir=False,
            size=meta.get("bytes"),
            modified_at=meta.get("uploadedAt"),
        )

    def _scope_missing(self) -> str:
        return f"Не задан scope вложений ({self.anchor}) в текущем ходе"

    async def _request(
        self, path: str, query: dict[str, str] | None, accept: str
    ) -> httpx.Response:
        url = f"{self._base_url}{self.api_base}{path}"
        headers = {"Accept": accept}
        token = self._bearer()
        if token:
            headers["Authorization"] = f"Bearer {token}"
        if self._client is not None:
            return await self._client.get(url, params=query or {}, headers=headers)
        async with httpx.AsyncClient() as client:
            return await client.get(url, params=query or {}, headers=headers)

    async def _api_json(self, path: str, query: dict[str, str] | None = None) -> dict[str, Any]:
        resp = await self._request(path, query, "application/json")
        if resp.status_code >= 400:
            raise RuntimeError(f"chat-backend {self.api_base}{path} -> HTTP {resp.status_code}")
        return resp.json()

    async def _api_bytes(
        self, path: str, query: dict[str, str] | None = None
    ) -> tuple[bytes, str | None]:
        resp = await self._request(path, query, "application/octet-stream")
        if resp.status_code >= 400:
            raise RuntimeError(f"chat-backend {self.api_base}{path} -> HTTP {resp.status_code}")
        return resp.content, resp.headers.get("content-type")


class ChatAttachmentsStoreBackend(AttachmentsStoreBackendBase):
    """Эфемерные вложения чата (Redis TTL), якорь ``/chat-attachments/``. Namespace — contextId.

    ``read_raw`` отдаёт БАЙТЫ оригинала (Фаза 4) — для детерминированного парсинга xlsx/csv.
    """

    anchor = "chat-attachments"
    api_base = "/api/chat-attachments"

    def __init__(
        self,
        *,
        base_url: str,
        context_id: Callable[[], str | None],
        bearer: BearerFn | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        super().__init__(base_url=base_url, bearer=bearer, http_client=http_client)
        self._context_id = context_id

    def _scope_for_manifest(self) -> dict[str, str] | None:
        context_id = self._context_id()
        return {"contextId": context_id} if context_id else None

    def _scope_for_file(self) -> dict[str, str] | None:
        return self._scope_for_manifest()

    async def read_raw(self, file_path: str) -> ReadRawResult:
        file_id = self._parse(file_path)
        if not file_id:
            return ReadRawResult(error=f"Неизвестный путь для raw: {file_path}")
        scope = self._scope_for_file()
        if scope is None:
            return ReadRawResult(error=self._scope_missing())
        try:
            content, mime = await self._api_bytes(f"/{quote(file_id)}/raw", scope)
            return ReadRawResult(content=content, mime_type=mime)
        except Exception as exc:  # noqa: BLE001
            return ReadRawResult(error=str(exc))


class ProjectAttachmentsStoreBackend(AttachmentsStoreBackendBase):
    """Durable-вложения проекта (Postgres), якорь ``/project-attachments/``. Scope — projectId.

    ``read_raw`` — read-only (durable/S3 raw = follow-up к Фазе 4).
    """

    anchor = "project-attachments"
    api_base = "/api/project-attachments"

    def __init__(
        self,
        *,
        base_url: str,
        project_id: Callable[[], str | None],
        bearer: BearerFn | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        super().__init__(base_url=base_url, bearer=bearer, http_client=http_client)
        self._project_id = project_id

    def _scope_for_manifest(self) -> dict[str, str] | None:
        project_id = self._project_id()
        return {"projectId": project_id} if project_id else None

    def _scope_for_file(self) -> dict[str, str] | None:
        # read по fileId не требует projectId — chat-backend резолвит в скоупе владельца.
        return {}


def _render_manifest(anchor: str, attachments: list[dict[str, Any]]) -> str:
    lines = [f"# Вложения ({anchor})", ""]
    for meta in attachments:
        flags = " _(большой — грепай, не читай целиком)_" if meta.get("isLarge") else ""
        lines.append(f"- **{meta.get('sourceName')}** — `/{anchor}/{meta.get('fileId')}`{flags}")
        if meta.get("summary"):
            lines.append(f"  - {meta['summary']}")
    if not attachments:
        lines.append("_нет вложений_")
    return "\n".join(lines)


def _one_line(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()[:200]

File: packages/ts-host/src/store-backend/attachments-store-backend.ts
import { currentBearer } from '../als'
import type {
  EditResult,
  FileInfo,
  GlobResult,
  GrepMatch,
  GrepResult,
  LsResult,
  ReadRawResult,
  ReadResult,
  StoreBackend,
  WriteResult,
} from './types'

/** DTO REST вложений chat-backend (см. attachments.types на стороне chat-backend). */
interface AttachmentMetaDto {
  fileId: string
  sourceName: string
  mime: string
  bytes: number
  sha256: string
  summary: string
  isLarge: boolean
  uploadedAt: string
  expiresAt?: string
}
interface AttachmentGrepHitDto {
  fileId: string
  sourceName: string
  line: number
  snippet: string
}

const READ_ONLY = 'Вложения доступны только для чтения (агенты read-only к file:<id>)'

/**
 * База для StoreBackend'ов вложений: тонкий HTTP-клиент к REST chat-backend (как ChatStoreBackend).
 * Сами файлы (Redis/Postgres) знает только chat-backend — единая точка auth/tenancy. Агент работает
 * штатной ФС-эргономикой deepagents `CompositeBackend`; физически операции уходят в chat-backend с JWT.
 *
 * MOUNT-RELATIVE по контракту CompositeBackend: composite срезает префикс маунта на входе и добавляет
 * его к путям результатов на выходе, поэтому бэкенд оперирует путями ОТНОСИТЕЛЬНО точки монтирования
 * и не знает (и не должен знать), куда смонтирован — один инстанс можно монтировать в несколько
 * префиксов. `anchor` — лишь ИМЯ канонического маунта (`/<anchor>/`): используется в текстах ошибок
 * и в markdown-манифесте (текст для LLM, которая видит внешние — смонтированные — пути).
 *
 * Виртуальная ФС (пути относительные):
 * - `/`          — манифест файлов: `ls` (структурно) / `read` (markdown с source_name/summary);
 * - `/{fileId}`  — `read` окна markdown (offset/limit);
 * - `grep` — серверный поиск по содержимому; `glob` — по имени файла; `write/edit` — read-only.
 */
abstract class AttachmentsStoreBackendBase implements StoreBackend {
  protected readonly baseUrl: string
  protected readonly bearer: () => string | undefined
  protected readonly fetchImpl: typeof fetch

  /** Сегмент-якорь пути (`chat-attachments` | `project-attachments`). */
  protected abstract readonly anchor: string
  /** База REST в chat-backend (`/api/chat-attachments` | `/api/project-attachments`). */
  protected abstract readonly apiBase: string
  /** Query для манифест-операций (ls/glob/grep). null → скоуп не резолвится (вернуть ошибку). */
  protected abstract scopeForManifest(): Record<string, string> | null
  /** Query для пофайловых операций (read content). chat → contextId; project → пусто. */
  protected abstract scopeForFile(): Record<string, string> | null

  constructor(opts: { baseUrl: string; bearer?: () => string | undefined; fetchImpl?: typeof fetch }) {
    this.baseUrl = opts.baseUrl.replace(/\/+$/, '')
    this.bearer = opts.bearer ?? currentBearer
    this.fetchImpl = opts.fetchImpl ?? globalThis.fetch
  }

  // ── ls ─────────────────────────────────────────────────────────────────────
  async ls(path: string): Promise<LsResult> {
    const fileId = this.parse(path)
    if (fileId === null) return { error: `Не директория: ${path}` }
    if (fileId) return { error: `Не директория: ${path}` }
    const scope = this.scopeForManifest()
    if (!scope) return { error: this.scopeMissing() }
    try {
      const { attachments } = await this.api<{ attachments: AttachmentMetaDto[] }>('/', scope)
      return { files: attachments.map((a) => this.fileInfo(a)) }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── read (директория → манифест; файл → окно markdown) ────────────────────────
  async read(path: string, offset?: number, limit?: number): Promise<ReadResult> {
    const fileId = this.parse(path)
    if (fileId === null) return { error: `Неизвестный путь: ${path}` }
    try {
      if (!fileId) {
        const scope = this.scopeForManifest()
        if (!scope) return { error: this.scopeMissing() }
        const { attachments } = await this.api<{ attachments: AttachmentMetaDto[] }>('/', scope)
        return { content: renderManifest(this.anchor, attachments), mimeType: 'text/markdown' }
      }
      const scope = this.scopeForFile()
      if (!scope) return { error: this.scopeMissing() }
      const query: Record<string, string> = { ...scope }
      if (offset !== undefined) query.offset = String(offset)
      if (limit !== undefined) query.limit = String(limit)
      const { content } = await this.api<{ content: string }>(`/${enc(fileId)}/content`, query)
      return { content, mimeType: 'text/markdown' }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── glob (по имени файла, клиентский фильтр манифеста) ────────────────────────
  // path игнорируем: scope (contextId/projectId) берётся из резолвера хода, не из пути.
  async glob(pattern: string, _path?: string): Promise<GlobResult> {
    const scope = this.scopeForManifest()
    if (!scope) return { error: this.scopeMissing() }
    try {
      const needle = stripGlob(pattern).toLowerCase()
      const { attachments } = await this.api<{ attachments: AttachmentMetaDto[] }>('/', scope)
      const files = attachments
        .filter((a) => !needle || a.sourceName.toLowerCase().includes(needle))
        .map((a) => this.fileInfo(a))
      return { files }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── grep (по содержимому, серверный поиск) ────────────────────────────────────
  async grep(pattern: string, _path?: string | null, _glob?: string | null): Promise<GrepResult> {
    const scope = this.scopeForManifest()
    if (!scope) return { error: this.scopeMissing() }
    try {
      const { matches } = await this.api<{ matches: AttachmentGrepHitDto[] }>('/search', {
        ...scope,
        q: pattern,
      })
      return {
        matches: matches.map<GrepMatch>((h) => ({
          // Путь относительный — внешний префикс добавит CompositeBackend.
          path: `/${h.fileId}`,
          line: h.line,
          text: `[${h.sourceName}] ${oneLine(h.snippet)}`,
        })),
      }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── read-only ────────────────────────────────────────────────────────────────
  write(): Promise<WriteResult> {
    return Promise.resolve({ error: READ_ONLY })
  }
  edit(): Promise<EditResult> {
    return Promise.resolve({ error: READ_ONLY })
  }
  // Вложения — markdown-ТЕКСТ; raw-бинарь не отдаём (агент читает окнами через `read`).
  readRaw(): Promise<ReadRawResult> {
    return Promise.resolve({ error: 'readRaw не поддерживается: используйте read (markdown-текст)' })
  }

  // ── helpers ────────────────────────────────────────────────────────────────
  /**
   * fileId сегмента (string), '' для корня-директории, null если путь не распознан.
   * Путь относителен точки монтирования (контракт CompositeBackend: префикс срезан до нас):
   * `/` → директория-манифест, `/<fileId>` → файл, глубже одного сегмента — не наш путь.
   * Якорной формы (`/chat-attachments/<id>`) больше нет: бэкенд не знает своего маунта.
   */
  protected parse(path: string): string | null {
    const seg = path.split('/').filter(Boolean)
    if (seg.length === 0) return ''
    if (seg.length === 1) return seg[0]
    return null
  }

  // Путь относительный — внешний префикс добавит CompositeBackend.
  protected fileInfo(a: AttachmentMetaDto): FileInfo {
    return { path: `/${a.fileId}`, is_dir: false, size: a.bytes, modified_at: a.uploadedAt }
  }

  protected scopeMissing(): string {
    return `Не задан scope вложений (${this.anchor}) в текущем ходе`
  }

  protected async api<T>(path: string, query?: Record<string, string>): Promise<T> {
    const url = new URL(`${this.baseUrl}${this.apiBase}${path}`)
    if (query) for (const [k, v] of Object.entries(query)) url.searchParams.set(k, v)
    const headers: Record<string, string> = { Accept: 'application/json' }
    const token = this.bearer()
    if (token) headers.Authorization = `Bearer ${token}`
    const res = await this.fetchImpl(url.toString(), { headers })
    if (!res.ok) throw new Error(`chat-backend ${this.apiBase}${path} → HTTP ${res.status}`)
    return res.json() as Promise<T>
  }
}

export interface ChatAttachmentsStoreBackendOptions {
  /** База REST chat-backend (тот же хост, что history-API). */
  baseUrl: string
  /** user-JWT для форварда. По умолчанию `currentBearer` из request-scope. */
  bearer?: () => string | undefined
  /** fetch (по умолчанию глобальный). */
  fetchImpl?: typeof fetch
  /** contextId текущего хода — namespace треда в Redis (в chat-backend: turn-scope). */
  contextId: () => string | undefined
}

/**
 * Эфемерные вложения чата (Redis TTL), монтируется на `/chat-attachments/`. Namespace треда —
 * contextId хода (резолвер из request-scope), tenant/owner chat-backend берёт из JWT.
 *
 * ```ts
 * new CompositeBackend(new StateBackend(), {
 *   "/chat-attachments/": new ChatAttachmentsStoreBackend({ baseUrl, contextId: currentTurnContextId }),
 * })
 * // агент: ls('/chat-attachments/'), read('/chat-attachments/<fileId>', 0, 200),
 * //        grep('лифты','/chat-attachments/')
 * ```
 */
export class ChatAttachmentsStoreBackend extends AttachmentsStoreBackendBase {
  protected readonly anchor = 'chat-attachments'
  protected readonly apiBase = '/api/chat-attachments'
  private readonly contextId: () => string | undefined

  constructor(opts: ChatAttachmentsStoreBackendOptions) {
    super(opts)
    this.contextId = opts.contextId
  }

  protected scopeForManifest(): Record<string, string> | null {
    const contextId = this.contextId()
    return contextId ? { contextId } : null
  }
  // Пофайловые операции тоже в namespace треда → нужен contextId.
  protected scopeForFile(): Record<string, string> | null {
    return this.scopeForManifest()
  }
}

export interface ProjectAttachmentsStoreBackendOptions {
  baseUrl: string
  bearer?: () => string | undefined
  fetchImpl?: typeof fetch
  /** projectId/slug текущего треда (в chat-backend выводится из Thread.projectId). */
  projectId: () => string | undefined
}

/**
 * Durable-вложения проекта (Postgres), монтируется на `/project-attachments/`. Манифест/grep скоупятся
 * по projectId; пофайловые read резолвятся по fileId в скоупе владельца (projectId не нужен).
 *
 * ```ts
 * new CompositeBackend(new StateBackend(), {
 *   "/project-attachments/": new ProjectAttachmentsStoreBackend({ baseUrl, projectId: currentTurnProjectId }),
 * })
 * ```
 */
export class ProjectAttachmentsStoreBackend extends AttachmentsStoreBackendBase {
  protected readonly anchor = 'project-attachments'
  protected readonly apiBase = '/api/project-attachments'
  private readonly projectId: () => string | undefined

  constructor(opts: ProjectAttachmentsStoreBackendOptions) {
    super(opts)
    this.projectId = opts.projectId
  }

  protected scopeForManifest(): Record<string, string> | null {
    const projectId = this.projectId()
    return projectId ? { projectId } : null
  }
  // read по fileId не требует projectId — chat-backend резолвит в скоупе владельца.
  protected scopeForFile(): Record<string, string> | null {
    return {}
  }
}

// ── helpers ─────────────────────────────────────────────────────────────────────
// Манифест — ТЕКСТ для LLM, которая видит внешние (смонтированные) пути; anchor здесь — имя
// канонического маунта `/<anchor>/`, а не знание бэкенда о фактической точке монтирования.
function renderManifest(anchor: string, attachments: AttachmentMetaDto[]): string {
  const lines = [`# Вложения (${anchor})`, '']
  for (const a of attachments) {
    const flags = a.isLarge ? ' _(большой — грепай, не читай целиком)_' : ''
    lines.push(`- **${a.sourceName}** — \`/${anchor}/${a.fileId}\`${flags}`)
    if (a.summary) lines.push(`  - ${a.summary}`)
  }
  if (attachments.length === 0) lines.push('_нет вложений_')
  return lines.join('\n')
}
function enc(s: string): string {
  return encodeURIComponent(s)
}
function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e)
}
function oneLine(s: string): string {
  return s.replace(/\s+/g, ' ').trim().slice(0, 200)
}
function stripGlob(pattern: string): string {
  return pattern.replace(/[*?]/g, '').trim()
}
